README.md
# OTIMCON

Open TIMing CONtrol is a hardware part of a sport timing
system. The system consists of controls (stations) which have RFID
readers/writers, while the competitors wear RFID chips.


### Description
OTIMCON hardware consists of:
- Arduino board (currently only ATmega 328 based)
- DS3231 RTC board
- MRFC522 RFID board


It is loosely based on popular orienteering systems, and it's best used for orienteering, trekking, and other sports in which timing on several controls are needed.

This is done as the price of RFID modules went down, good RTC is easy to find, and microcontroller is just a glue. It's open so other makers can upgrade the system, or add additional modules like communication over WiFi / Bluetooth, a small screen with information, build a permanent course where the board is powered over solar cells etc.


### Usage of the system
1. before start a CLEAR control (station) is used to clear out the card and prepare it
2. on start an additional station can be used (with control ID 251). On mass start a prewritten control with time can be added (WRITE CONTROL command), or a control with frozen time (FROZEN_CONTROL) can be used.
3. on the racetrack the competitors touch the RFID card by the control, where each control has its unique id (from 1 to 250)
4. on the finish the competitors touch against the finish control (ID 252).
5. after the finish the station in READOUT mode is used to read the timing on each control.

Before step 1. an administrator can reset the backup memory of each control so the memory of which cards passed can be saved.

### Usage of the station
Stations support serial communication (38400 8 N 1). All configuration is done through serial terminal. Use "?" for help.
_Note_ : use Arduino serial monitor for communication or a system that sends _newline_ at the end of the line, as most systems don't.
_Note_ : baud rate of 38400 is used. On some Arduino boards such speed isn't supported and it needs to be slowed - suggested speed is 9600.

### Serial commands
* __SET__
	* TIME YYYYMMDDhhmmss : Set time in year[Y], month[M], day[D], hour[h], minute[m], second[s] format  (eg. 20160817120024 is 2016/08/17 12:00:24)
	* CTRL number         : Set control ID as number (1-250), with special codes for START and FINISH stations (special id number defined in code)
	* MODE type           : Set modes (possible modes are CLEAR, CONTROL, CONTROL_WITH_READOUT, FROZEN_CONTROL, FULL_READOUT, PRINT, READOUT, WRITER)
	* RESET_BACKUP        : Reset pointer to 0th location on external EEPROM backup memory. In reality not neccessary as it rolls over, but nice to have a clean start.


* __GET__
	* TIME                : Get time in format YYYY/MM/DD hh:mm:ss - year[Y], month[M], day[D], hour[h], minute[m], second[s] eg. 2016/08/17 12:00:24
	* CTRL                : Get control ID as number (1-250 for normal controls, special codes for START and FINISH)
	* MODE                : Get mode status (CLEAR, CONTROL, CONTROL_WITH_READOUT, FROZEN_CONTROL, FULL_READOUT, PRINT, READOUT, WRITER)
	* BACKUP              : Get backup of whole external EEPROM (backup) memory. It returns all locations, even those which are not used.
	* VERSION             : Get current firmware version
	* VOLTAGE             : Get current battery voltage if the battery is connected to analog port 2. (may not work as expected)

  * __WRITE__   
  	* INFO  Any_text             : Write any text up to 30 characters which will be used in PRINT command as a CARD owner/user
  	* CTRL number YYYYMMDDhhmmss : Writes a control at the end of current controls. Useful to add a finish time in case of error, or to setup some controls (eg. START) prior to race.
  _Note_ : WRITE commands work only in WRITER mode.

  * __PING__
    * command PING will respond with PONG. It should be used as a keep awake system.


### Questions:
* What are CLEAR, CONTROL, CONTROL_WITH_READOUT, FROZEN_CONTROL, READOUT, FULL_READOUT and PRINT controls?
	* CONTROL is a normal station which saves data to RFID card, and sends (via serial) the UID of the card, timestamp in UNIX epoch format and a readable time.
  * CLEAR completely erases all the data on the RFID card.
  * READOUT reads every written control on the RFID card, from the first written to the last written on the card, and sends it via serial. It should be used with software on a computer.
  * CONTROL_WITH_READOUT does the same as control, but additionaly reads every previous control off the card. This is usable if some of the controls in the middle of the race are connected over Internet. Also it's usable to have one control for FINAL + automatic readout
  * FULL_READOUT is used when all 125 places on the card want to be read.
  * FROZEN_CONTROL is a station which has it's clock stopped. So every competitor will have the same time on that control. It's function is primarily for races with mass start (all competitors start at the same time), so the organizers do not need to set that into the software.
  * PRINT is a station which will nicely format and calculate the results of the station on the card, and send it via serial. It also prints the name of the user/owner of the card. It doesn't need the software on a computer.
_Note_ : There is a WRITER mode, but in that mode a command is to be given to a station prior of putting a card next to the RFID reader.


* How many controls can one RFID card accommodate?
	* Short answer: 125
	* Long answer: The system is prepared for MiFare 1K cards, which have 64 blocks (divided in 16 sectors) of 16 bytes. Each sector has 3 usable blocks, and one trailer block which has special meaning. First sector (blocks 0 - 3)is also off the limits. Second sector is prepared for info block (block 4) while blocks 5 and 6 are left free for anything (eg. competitor names). Data starts at block 8. Out of 56 blocks, there are 14 trailer blocks which are unusable, so control data can come to 42 blocks. Data about the control is saved in 5 bytes: 1 byte is control ID (number between 0-255), and 4 bytes is UNIX timestamp of the coming to control. So there are 3 controls per block - in total 42 x 3 = 126. There is a hard limit in code to 125.
	* For an visual answer look at the [CardAllocation](https://github.com/stajp/otimcon/blob/master/CardAllocationOTIMCON.ods)


* Is data safe on RFID card?
	* Data retention of the RFID memory is 10 years, or 200000 writes. So it's OK for a normal use. Additionaly, every block on the card has an additional CRC8 check on the end, so the data is written as it should.


* For what is BACKUP used?
	* If there is support for external EEPROM, then after writing data about the control onto the RFID chip, there is an additional write to external EEPROM with UID of the RFID card and timestamp.


* How big is BACKUP?
	* Backup saves last 512 cards (one card can be more than once, for those who forget to erase).

* What is info block in the RFID card?
	* Info block is block 4, where the last location is saved, last control and time of the last control.

* It looks like the serial port isn't working?
	* Make sure you're using a serial terminal which sends \n newline (most of the terminal software doesn't). Serial monitor in Arduino IDE works nicely.
	* Also make sure you're not using LOW POWER mode - in that case use an RFID card on the station prior to serial communication (see next question).

  * My station worked fine when I turned it on, but after an hour they are dead?
    * Check that LOW_POWER option is not active, and that HIGH_POWER is active.
    * If LOW_POWER is option is active, then each station goes through three states of work which are AWAKE -> SHALLOW SLEEP -> DEEP SLEEP.
      The station has to be in AWAKE mode for normal function (read/write), and the system is working at full speed.
      If no card or no serial command is present for 5 seconds, the station goes into SHALLOW SLEEP. In that mode station checks for card or serial command once every 250ms. If a card or serial command is present, it goes into AWAKE state.
      If no card or serial command is present in an hour of work in SHALLOW SLEEP, the station goes into DEEP SLEEP. In DEEP SLEEP the card or serial command is checked once every 2 seconds. Due to other electronics which needs to wake up, it can be up to 10 seconds needed for a control to fully wake up into AWAKE state. Just be patient.
      Time of 5 seconds used in AWAKE is chosen to not slow runners which came in bursts (several runners at once). Station can be forced in AWAKE state if any of the serial command is issued - PING can be used for keep AWAKE (or use HIGH_POWER option).
      The time of SHALLOW SLEEP of 250ms is not noticeable to runners. It can be lowered if needed.

* How can you use a MRFC522 on 5V?
	* MRFC522 is powered on 3.3V, and it will burn if you try to power it on 5V. But it seems that it can live with 5V on it's SPI port. The schematic given is my test bed, and all the orhers are done on a 3.3V Pro mini. DS3231 can work on either 3.3V or 5V, so it doesn't matter to it, this is just for the MRFC522.


* I did a RESET_BACKUP function but old data is still present in backup?
	* RESET_BACKUP just resets the internal counter back to 0 (first position), and starts overwriting from there. Nothing to be alarmed, it's like that by design.


* Is this code production ready, and what's the license?
  * The code is tested as much as authors was able. It's (very) possible there are some bugs. The earlier versions of the system are in use by several early adopters, so the basic CONROL-READOUT-CLEAR does work (very well).
  * The license is GNU General Public License v3.0 , so you  can do whatever you want with this, but
  1. the author is not liable or has to give anything more to the users than what's given.
  2. any further changes to the code must be given away, under the same license.


documentation/CHANGELOG
History:
 * - version 0.1 20160601
 *      First version is based on
 *      - ideas about OPORCON ( my project based on 1-wire iButtons and using PIC16F628, available at http://stipe.predanic.com/?Projekti/OPORCON )
 *      - MFRC example by Rudy Schlaf for www.makecourse.com http://makecourse.weebly.com/week10segment1.html  -> READ IT, WATCH THE VIDEO!!!!!!
 *     Rudely hacked to prove that it works.
 *
 * - version 0.11 20160605
 *      - cleaned up the functions needed for this to work
 *
 * - version 0.12 20160816
 *      - further cleaning, added read back to check if everything is written as it should, added the bleep and basic sleep
 *      - a lot of ifdef preprocessing added, to make everything clean
 *
 * - version 0.13 20160816
 *      - added support for readout and clear (NOT TESTED YET!)
 *
 * - version 0.14 20160816
 *      - added support for serial CLI for setting and getting data
 *      - support for different modes and control ID's done through CLI
 *      - changed name from OTICON to OTIMCON for search engines
 *
 * - version 0.15 20160817
 *      - CLI tested
 *      - added the AD conversion for battery measurement (not tested)
 *      - modes and control saved into ATMega328 EEPROM
 *
 * - version 0.16 201619
 *      - make backup of card UID and time to EEPROM on Ebay DS3231 boards
 *
 * - version 0.17 201619
 *      - sleep for low power work -> done it poorly, should revisit one day
 *         -> done only for microcontroller, RFID and RTC should also go to sleep
 *         -> low power now puts the microcontroller to sleep for 125ms, wakes up, checks the RFID and goes back to sleep.
 *            after reading a card the system doesn't go to any kind of sleep for half a minute(!!!) and serial works as expected (also the reading of the cards is quicker!)
 *            --> if serial is used, that resets the counter for another minute.
 *         -> if low power is not used (comment out a define), then serial always works as expected.
 *         -> by default LOW POWER is NOT USED!
 *
 *  - version 0.18 201700306
 *      - sleep for low power works doesn't block the system, but it's not tested with multimeter.
 *      - Low power is now DEFAULT.
 *      - changed the identifier of a card ready for OTIMCON from "1" to custom number defined in  OTIMCON_CARD_TYPE
 *      - changed the sound and led signalling. Now it works better.
 *      - added mode CRAZY_READOUT. It will read the whole RFID card, ignoring the START and FINISH controls.
 *      - fixed a bug with position 0
 *      - fixed a bug when card was removed in a long readout
 *
 *
 *  - version 0.19 201700311 (wasn't pushed to Github)
 *      - sleep for low power works - tested with multimeter.
 *        * The system based on _regular_ Arduino board, MRFC522 (connected to 3.3V on Arduino board) and RTC module, all powered up constantly - eats up 25mA when sleeping, 55mA when working.
 *          -> power (5V from powerbank) was connected to 5V pin on Arduino, and current was measured on multimeter
 *          -> RTC&EEPROM module 5mA,
 *          -> MRFC522 takes 10mA when working, 2mA when sleeping. Power LED is still on board, it's probably it, should try desoldering it.
 *          -> the rest (aroung 18mA sleeping, 40mA working) is Arduino board - but it has to power USB-serial converter, 3.3V regulator, power LED, and something is lost on the 5V regulator.
 *            It's as expected, as Nick Gammon (link below) has similar readings for regular Arduino board.
 *          - for testing: Pro mini, Pro mini with power LED desoldered, desoldering power LED on MRFC522
 *
 *      - decided to add option to power the RTC and EEPROM through Arduino pin 5 (setup RTC_POWER_PIN for other). It can still be powered constantly, but it draws 5mA doing nothing,
 *        Don't worry about the correct time, there is an RTC backup battery (ideas by https://edwardmallon.wordpress.com/2014/05/21/using-a-cheap-3-ds3231-rtc-at24c32-eeprom-from-ebay/
 *        and https://www.gammon.com.au/forum/?id=11497
 *
 *  - version 0.2 20170327
 *     - changed how READOUT works, it reads from start of the card till the last written location on card (previously looked for START or previuos FINISH and started from there)
 *     - changed the READOUT output format
 *     - CONTROL_WITH_READOUT from now on also reads from start of the card (same as READOUT). It takes a lot of time to read the card!
 *     - added PRINT mode which calculates and prints time from START to FINISH (starts from the last written control and goes towards the beginning looking for START)
 *     - support for FULL_READOUT which reads the whole card (all 125 locations)
 *     - increased the serial speed to 38400, should work OK if good crystal is used. Change to 9600 if internal oscillator is used.
 *     - added a PING command which responds with PONG. Should be used as keepalive messaging.
 *     - added the WRITE INFO and WRITE CONTROL options
 *     - added detection of START and FINISH control numbers in both input and output
 *     - added a high power mode for use with auto shut off powerbanks (Chris Gkikas equipment).
 *     - fixed an anoying turn off bug with wrong time if the board was just turned off, and not reprogrammed.
 *     - added FROZEN_CONTROL which is a standard control but with a fixed time. This is for setting start time for mass start (Note - any control can be set, not only start)!
 *     - cleaned the function which does CLEAR
 *
 * - version 0.21 20180331
 *     - fixed detection of card which isn't removed after the beep, to stop multiple prints, reads and writes.
 *     - changed organization of files
 *
 * - version 0.22 20180402
 *     - fixed organization of files
 *     - added bleep as command (not documented in help)
 *     - Python 2.7 command line scripts for easier work with OTIMCON and for printing tickets


scripts/ticket_print.py

# Uses LPR default printer for printing (presumably OTIMCON tickets).
# LPR is used for printing on linux and macOS.
# For Windows check startCommand below.
#
# Usage: ticket_print -p PORT -s speed
# eg. python ticket_print.py -p /dev/cu.myPort -s 9600
#
# use python ticker_print.py --help for more info
#

import glob
import getopt

# edit this to suit your OS, or use the "command" option when invoking this script
# linux and macOS  startCommand = 'lpr'
# windows          startCommand = 'print D:\\\MACHINE_NAME\PRINTER_NAME'
#                  (possible additional parameters needed, check number of '\')
startCommand = 'lpr'

def usage():
    print "Ticket printing for OTIMCON"
    print ""
    print "Usage:"
    print "  ticket_print -pPORT (-sSPEED) [-i] [-c'COMMAND']"
    print "  ticket_print --port=PORT (--speed=SPEED) [--increment] [--command='COMMAND']"
    print "  ticket_print -h | --help"
    print ""
    print "Options:"
    print "  -h --help       Show this screen"
    print "  -p --port       Serial port for OTIMCON"
    print "  -s --speed      Speed of serial port for OTIMCON, default 38400"
    print "  -i --increment  Save every ticket, with incremental filename"
    print "  -c --command    Command which starts the print, default 'lpr'"
    print ""
    print "Example:"
    print "  ticket_print -p/dev/cu.wchusbserial640 -s9600 -i"
    print "  ticket_print --port=COM4 --speed=38400 --command='print /D:\\\\\MACHINE\PRINTER"
    print ""
    print "Procedure:"
    print "  1. Connect OTIMCON station, look up the serial port."
    print "  2. Set receipt printer as default printer, or set the --command option."
    print "  3. Start this script. It will automatically convert OTIMCON station to PRINT mode!"
    print "  4. Use the script. Exit with CTRL-C."
    print ""
    print "Copyright (c) Stipe Predanic, given under GNU General Public License v3.0"

def inputParse():
    global startCommand
    try:
        opts, args = getopt.getopt(sys.argv[1:], "hp:s:c:i", ["help", "port=", "speed=", "command=", "increment"])
    except getopt.GetoptError as err:
        # print help information and exit:
        print str(err)  # will print something like "option -a not recognized"
        usage()
        sys.exit(2)
    port = ''
    speed = '38400'
    increment = False
    for o, a in opts:
        if o in ("-p", "--port"):
            port = a
        elif o in ("-s", "--speed"):
            speed = a
        elif o in ("-i", "--increment"):
            increment = True
        elif o in ("-h", "--help"):
            usage()
            sys.exit()
        elif o in ("-c", "--command"):
            startCommand = a
        else:
            assert False, "unhandled option"
    return port, speed, increment

def convertToPrintMode(ser):
    print "\n!!! NOTICE: Converting the OTIMCON to PRINT mode !!!"
    print "Press CTRL-C in 3 seconds to stop this."
    print ""
    time.sleep(3)
    readData = ser.readline()
    if readData[0:11] == 'Starting...':
        readData = ser.readline()
        print "Connected..."
        if readData[0] == '>':
            ser.flush()
            ser.write(b"SET MODE PRINT\n")
            time.sleep(0.5)
            print "Setting mode PRINT..."
            readData = ser.readline()
            readData = ser.readline()
            if readData[0:10] == 'Mode:PRINT':
                readData = ser.readline() # catch the last ">"
                return True
            else:
                print "!!! Error: OTIMCON refuses PRINT mode!"
        else:
            print "!!! Error: OTIMCON doesn't behave as expected!"
    else:
        print "!!! Error: Cannot connect to OTIMCON. Check port & speed!"
    return False

if __name__ == '__main__':
    import sys, subprocess, serial, time

    port, speed, incrementFilename = inputParse()
    if port is '':
        print "\n!!! Error: Port must be defined!"
        print "Use ticket_print --help for usage.\n"
        sys.exit(2)
    try:
        ser = serial.Serial(port, speed, timeout=1)
    except Exception as err:
        print "\n!!! Error: Cannot connect to serial port " + port
        print str(err)
        print ""
        sys.exit(2)

    print "\nConnected to port " + port + " at " + speed + " baud."
    print "Printing using '" + startCommand +"' command."

    # convert OTIMCON to PRINT mode
    if convertToPrintMode(ser) is False:
        print "\n!!! Error: Cannot convert OTIMCON to PRINT mode."
        print "OTIMCON state is unknown, check the current OTIMCON mode manually.\n"
        sys.exit(2)

    print ""
    print "Ready, waiting for OTIMCON cards..."

    # everythin OK, loop forever
    ticketNo = 1
    newData = False
    while True:
        time.sleep(0.02)  # sleep added so there is always something is in serial buffer until printing is really needed
        readData = ser.readline()

        # if data is not empty, check if it's the first, if it is - open file, write it, otherwise just append
        if readData is not '':
            if newData is False:
                if incrementFilename is True:
                    print "Incrementing filename number.."
                    fileName = "temp_ticket_" + str(ticketNo)
                else:
                    fileName = "temp_ticket"
                fileOutput = open(fileName, "w")
            newData = True
            sys.stdout.write(readData)
            fileOutput.write(readData)
        else: # no new data in buffer, close the temporaty ticket file and print it
            if newData:
                fileOutput.close()
                print "File closed, printing the " + str(ticketNo) + ". ticket using the '" + startCommand + "' command."
                time.sleep(0.1)
                subprocess.call(startCommand + " " + fileName, shell=True)
                ticketNo += 1
                newData = False


scripts/otimcon_setup.py

# Command line setup for OTIMCON

# Note: Does _NOT_ support all OTIMCON options!

import glob
import getopt

def usage():
    print "Command line setup for OTIMCON"
    print ""
    print "Usage:"
    print "  otimcon_setup -pPORT (-sSPEED) [-g] [-t]"
    print "  otimcon_setup --port=PORT (--speed=SPEED) [--get] [--settime] [--setOPTIONS=setValue]"
    print "  otimcon_setup -h | --help"
    print ""
    print "Options:"
    print "  -h --help       Show this screen"
    print "  -p --port       Serial port for OTIMCON"
    print "  -s --speed      Speed of serial port for OTIMCON, default 38400"
    print "  -g --get        Get all data about OTIMCON box"
    print "  -t --settime    Sets the current time taken from the PC"
    print "     --setmode    Sets the OTIMCON to mode (CLEAR|CONTROL|PRINT|READOUT)"
    print "     --setctrl    Sets the number of control (1-250|START|FINISH)"
    print "     --writeinfo  Sets the OTIMCON to WRITER mode and writes info to card"
    print ""
    print "Example:"
    print "  otimcon_setup -p/dev/cu.wchusbserial640 -s9600 -t --setmode=CONTROL --setctrl=START"
    print "  otimcon_setup --port=COM4 --speed=38400 --setmode=CLEAR"
    print "  otimcon_setup --port=/dev/dev/ttyUSB0 -s 9600 --settime --setmode=CONTROL --setctrl=45"
    print "  otimcon_setup -p /dev/dev/ttyUSB0 --speed=38400 --writeinfo='John Smith RED'"
    print "  otimcon_setup -p COM3 -s 38400 -g"
    print ""
    print "Procedure:"
    print "  1. Connect OTIMCON station, look up the serial port."
    print "  2. Start command line setup"
    print ""
    print "Notes:"
    print "  1. Modes CONTROL WITH PRINTOUT, FULL READOUT, FROZEN CONTROL are not supported through"
    print "     this software. Also WRITER mode where a control is written. Those can still be used"
    print "     through serial connection to OTIMCON."
    print "  2. Setting time is not necessary for every option, but is suggested to set time every"
    print "     time as possible."
    print "  3. Writing info leaves OTIMCON in WRITER mode."
    print "  4. Option 'get' will give every information, even those which are maybe not supported"
    print "     by hardware. It is not allowed to use get and set at the same time."
    print ""
    print "Copyright (c) Stipe Predanic, given under GNU General Public License v3.0"

def inputParse():
    try:
        opts, args = getopt.getopt(sys.argv[1:], "hgtp:s:", ["help", "port=", "speed=", "get", "settime", "setmode=", "setctrl=", "writeinfo="])
    except getopt.GetoptError as err:
        # print help information and exit:
        print str(err)  # will print something like "option -a not recognized"
        usage()
        sys.exit(2)
    port = ''
    speed = '38400'
    get = False
    setTime = False
    mode = ''
    ctrl = ''
    write = False
    writeInfo = ''
    for o, a in opts:
        if o in ("-p", "--port"):
            port = a
        elif o in ("-s", "--speed"):
            speed = a
        elif o in ("-g", "--get"):
            get = True
        elif o in ("-h", "--help"):
            usage()
            sys.exit()
        elif o in ("-t", "--settime"):
            setTime = True
        elif o in ("--setmode"):
            mode = a
        elif o in ("--setctrl"):
            ctrl = a
        elif o in ("--writeinfo"):
            write = True
            writeInfo = a
        else:
            assert False, "unhandled option"
    return port, speed, get, setTime, mode, ctrl, write, writeInfo

def wakeUp(ser):
    ser.write(b"\n")
    readData = ser.readline()
    counter = 0
    while counter < 6:
        readData = ser.readline()
        if readData[0:1] == '>':
            break;
        else:
            counter +=1;
        if counter == 3:
            ser.write(b"\n")
    counter = 0
    ser.write(b"\n")

    while counter < 6:
        readData = ser.readline()
        if readData[0:1] == '>':
            break;
        else:
            counter +=1;
        if counter == 3:
            ser.write(b"\n")

    time.sleep(1)
    ser.write(b"PING\n")
    time.sleep(0.5)
    readData = ser.readline()
    readData = ser.readline()

    if readData[0:4] == 'PONG':
        return True
    return False

def getMode(ser):
    print "Detecting current OTIMCON mode..."
    if wakeUp(ser) is True:
        ser.write(b"GET MODE\n")
        time.sleep(0.1)
        readData = ser.readline()
        readData = ser.readline()
        if readData[0:5] == 'Mode:':
            return readData[5:].rstrip()

def convertToMode(ser, mode):
    print "\n!!! NOTICE: Converting OTIMCON to " + mode + " mode !!!"
    print "Press CTRL-C in 3 seconds to stop this."
    print ""
    time.sleep(3)
    if wakeUp(ser) is True:
        readData = ser.readline()
        print "OTIMCON awake, connected..."
        if readData[0] == '>':
            ser.flush()
            ser.write(b"SET MODE " + mode + "\n")
            time.sleep(0.5)
            print "Setting mode " + mode + "..."
            readData = ser.readline()
            readData = ser.readline()
            expectedReponse='Mode:'+mode
            if readData[0:len(expectedReponse)] == expectedReponse:
                readData = ser.readline() # catch the last ">"
                return True
            else:
                print "!!! Error: OTIMCON refuses " + mode + " mode!"
        else:
            print "!!! Error: OTIMCON doesn't behave as expected!"
    else:
        print "!!! Error: Cannot connect to OTIMCON. Check port & speed!"

    return False

def convertToControlMode(ser, ctrl):
    currentMode = getMode(ser)
    if currentMode[0:len(mode)] != mode:
        result = convertToMode(ser,'CONTROL')
        if result is True:
            print "Success: OTIMCON in CONTROL mode"
    else:
        print "Success: OTIMCON was in CONTROL mode"
        result = True

    if ctrl == '':  # control number isn't necessary
        return True

    validCtrl = False
    if (ctrl == 'START' or ctrl == 'FINISH'):
        validCtrl = True
    elif (int(ctrl) > 0 and int(ctrl) < 256):
        validCtrl = True
    if (result is True and validCtrl is True):
        ser.write(b"SET CTRL " + ctrl + "\n")
        time.sleep(0.5)
        print "Setting control to " + ctrl + "..."
        readData = ser.readline()
        readData = ser.readline()
        expectedReponse='Control:'+ctrl
        print expectedReponse
        print readData
        if readData[0:len(expectedReponse)] == expectedReponse:
            readData = ser.readline() # catch the last ">"
            return True
        else:
            print "!!! Error: OTIMCON refuses to set control to " + ctrl + "!"
    else:
        print "!!! Error: OTIMCON isn't in CONTROL mode or invalid control designation"
    return False


def writeInfoToCard(ser, info):
    if (convertToMode(ser, 'WRITER') is True):
        ser.write(b"WRITE INFO " + info + "\n")
        time.sleep(0.1)
        readData = ser.readline()
        if readData[0:30] == "Waiting for card to write info":
           print "Put chip at the OTIMCON station"
           counter = 0
           while counter < 5:
               time.sleep(1)
               readData = ser.readline()
               if readData[0:24] == "WRITING INFO TO CARD: OK":
                   print "Succes: Info written to card."
                   return True
               else:
                   counter += 1
           print "!!! Error: info not written"
        else:
            print "!!! Error: OTIMCON refuses to write info."
    else:
        print "!!! Error: OTIMCON not set to WRITER mode."
    return False

def getAllData(ser):
    print "Trying to connect to OTIMCON..."
    if wakeUp(ser) is True:
        print "OTIMCON awake, connected..."
        print "\nOTIMCON DATA:\n---------------"
        ser.write(b"GET VERSION\n")
        time.sleep(0.1)
        readData = ser.readline()
        readData = ser.readline()
        print readData.rstrip()

        ser.write(b"GET TIME\n")
        time.sleep(0.1)
        readData = ser.readline()
        readData = ser.readline()
        print readData.rstrip()

        ser.write(b"GET MODE\n")
        time.sleep(0.1)
        readData = ser.readline()
        readData = ser.readline()
        print readData.rstrip()

        if readData[5:].rstrip() == 'CONTROL':
            ser.write(b"GET CTRL\n")
            time.sleep(0.1)
            readData = ser.readline()
            readData = ser.readline()
            print readData.rstrip()

        ser.write(b"GET VOLTAGE\n")
        time.sleep(0.1)
        readData = ser.readline()
        readData = ser.readline()
        print readData.rstrip()




if __name__ == '__main__':
    import sys, subprocess, serial, time, datetime

    port, speed, get, setTime, mode, ctrl, write, writeInfo = inputParse()
    if port is '':
        print "\n!!! Error: Port must be defined!"
        print "Use otimcon_setup --help for usage.\n"
        sys.exit(2)
    try:
        ser = serial.Serial(port, speed, timeout=1)
    except Exception as err:
        print "\n!!! Error: Cannot connect to serial port " + port
        print str(err)
        print ""
        sys.exit(2)

    print "\nConnected to port " + port + " at " + speed + " baud."

    if get is True:
        print "GET DATA" #TODO: remove
        getAllData(ser)
    else:
        if write is True:
            if writeInfoToCard(ser, writeInfo) is False:
                print "\n!!! Error: Cannot convert OTIMCON to WRITE mode and write info."
                print "OTIMCON state is unknown, check the current OTIMCON mode manually.\n"
                sys.exit(2)
            else:
                print "Success: OTIMCON wrote info on card."
        elif (mode[0:5] == 'PRINT') or (mode[0:5] == 'CLEAR') or (mode[0:7] == 'READOUT'):
            currentMode = getMode(ser)
            if currentMode[0:len(mode)] == mode:
                print "Success: OTIMCON was already in " + mode + " mode."
            else:
                print "Current mode is " + currentMode + ", will be changed into " + mode + "."
                if convertToMode(ser,mode) is False:
                    print "!!! Error: Cannot convert OTIMCON to " + mode + " mode."
                    print "OTIMCON state is unknown, check the current OTIMCON mode manually.\n"
                    sys.exit(2)
                else:
                    print "Success: OTIMCON is in " + mode + " mode."
        elif mode == 'CONTROL':
            print "SET CONTROL" #TODO: remove
            if convertToControlMode(ser, ctrl) is False:
                print "!!! Error: Error while setting OTIMCON to CONTROL mode or the control designation."
                print "OTIMCON state is unknown, check the current OTIMCON mode manually.\n"
                sys.exit(2)
            else:
                if ctrl != '':
                    print "Success: OTIMCON in CONTROL mode with designation " + ctrl + "."
                else:
                    print "Success: OTIMCON in CONTROL mode, use GET option for desigation."
        if setTime is True:
            print 'Setting time..'
            if wakeUp(ser) is True:
                print "OTIMCON awake, trying to set time..."
                now = datetime.datetime.now()
                ser.write(b"SET TIME " + now.strftime("%Y%m%d%H%M%S")+"\n")
                time.sleep(0.1)
                readData = ser.readline()
                readData = ser.readline()
                if readData[0:5] == 'Time:':
                    print "Time set, OTIMCON "+ readData
                else:
                    print "!!! Error: Time not set!"
            else:
                print "!!! Error: Not connected to set time."
    print ""
